README.md
# cluster-service

[![Build Status](https://travis-ci.org/godaddy/node-cluster-service.png)](https://travis-ci.org/godaddy/node-cluster-service) [![NPM version](https://badge.fury.io/js/cluster-service.png)](http://badge.fury.io/js/cluster-service) [![Dependency Status](https://gemnasium.com/godaddy/node-cluster-service.png)](https://gemnasium.com/godaddy/node-cluster-service) [![Bitdeli Badge](https://d2weczhvl823v0.cloudfront.net/godaddy/node-cluster-service/trend.png)](https://bitdeli.com/free "Bitdeli Badge")

[![NPM](https://nodei.co/npm/cluster-service.png?downloads=true&stars=true&downloadRank=true)](https://www.npmjs.org/package/cluster-service) [![NPM](https://nodei.co/npm-dl/cluster-service.png?height=2)](https://nodei.co/npm/cluster-service/)

## Install

	npm install cluster-service

https://npmjs.org/package/cluster-service



## About

	Turn your single process code into a fault-resilient, multi-process service with
	built-in REST & CLI support. Restart or hot upgrade your web servers with zero
	downtime or impact to clients.

Presentation:

http://x.co/bpnode

Video:

http://x.co/bpnodevid


## Getting Started

Your existing application, be it console app or service of some kind:

	// server.js
	console.log("Hello World");

Leveraging ```cluster-service``` without adding a line of code:

	npm install -g cluster-service
	cservice "server.js" --accessKey "lksjdf982734"
	// cserviced "server.js" --accessKey "lksjdf982734" // daemon

This can be done without a global install as well, by updating your ```package.json```:

	"scripts": {
		"start": "cservice server.js --accessKey lksjdf982734"
	},
	"dependencies": {
		"cluster-service": ">=0.5.0"
	}

Now we can leverage ```npm``` to find our local install of ```cluster-service```:

	npm start

Or, if you prefer to control ```cluster-service``` within your code, we've got you covered:

	// server.js
	require("cluster-service").start({ workers: "./worker.js", accessKey: "lksjdf982734" });

	// worker.js
	console.log("Hello World"); // notice we moved our original app logic to the worker



## Talk to it

Now that your service is resilient to worker failure, and utilizing all cores of your machine, lets talk to it.
With your service running, type into the command-line:

	restart all

or for a full list of commands...

	help

or for help on a specific command:

	help {command}

We can also issue commands from a seperate process, or even a remote machine (assuming proper access):

	npm install -g cluster-service
	cservice "restart all" --accessKey "my_access_key"

You can even pipe raw JSON for processing:

	cservice "restart all" --accessKey "my_access_key" --json

Check out ***Cluster Commands*** for more.



## Start Options

When initializing your service, you have a number of options available:

	cservice "server.js" --accessKey "123"

Or via JSON config:

	cservice "config.json"

Or within your node app:

	// server.js
	// inline options
	require("cluster-service").start({ workers: "worker.js", accessKey: "123" });
	// or via config
	require("cluster-service").start({ config: "config.json" });

### Options:

* `workers` - Path of worker to start. A string indicates a single worker,
  forked based on value of ```workerCount```. An object indicates one or more worker objects:
  ```{ "worker1": { worker: "worker1.js", cwd: process.cwd(), count: 2, restart: true } }```.
  This option is automatically set if run via command-line ```cservice "worker.js"``` if
  the ```.js``` extension is detected.
* `accessKey` - A secret key that must be specified if you wish to invoke commands from outside
  your process. Allows CLI & REST interfaces.
* `config` - A filename to the configuration to load. Useful to keep options from having to be inline.
  This option is automatically set if run via command-line ```cservice "config.json"``` if
  the ```.json``` extension is detected.
* `host` (default: "localhost") - Host to bind to for REST interface. (Will only bind if `accessKey`
  is provided)
* `port` (default: 11987) - Port to bind to. If you leverage more than one cluster-service on a
  machine, you'll want to assign unique ports. (Will only bind if accessKey is provided)
* `workerCount` (default: os.cpus().length) - Gives you control over the number of processes to
  run the same worker concurrently. Recommended to be 2 or more to improve availability. But some
  workers do not impact availability, such as task queues, and can be run as a single instance.
* `cli` (default: true) - Enable the command line interface. Can be disabled for background
  services, or test cases. Running `cserviced` will automatically disable the CLI.
* `ssl` - If provided, will bind using HTTPS by passing this object as the
  [TLS options](http://nodejs.org/api/tls.html#tls_tls_createserver_options_secureconnectionlistener).
* `run` - Ability to run a command, output result, and exit. This option is automatically
  set if run via command-line ```cservice "restart all"``` and no extension is detected.
* `json` (default: false) - If specified in conjunction with ```run```,
  will *only* output the result in JSON for
  consumption from other tasks/services. No other data will be output.
* `silent` (default: false) - If true, forked workers will not send their output to parent's stdio.
* `allowHttpGet` (default: false) - For development purposes, can be enabled for testing, but is
  not recommended otherwise.
* `restartOnMemUsage` (default: disabled) - If a worker process exceeds the specified memory threshold
  (in bytes), the process will be restarted gracefully. Only one worker will be restarted at a time.
* `restartOnUpTime` (default: disabled) - If a worker process exceeds the specified uptime threshold
  (in seconds), the process will be restarted gracefully. Only one worker will be restarted at a time.
* `restartConcurrencyRatio` (default `0.33`) - The ratio of workers that can be restarted concurrently
  during a restart or upgrade process. This can greatly improve the speed of restarts for applications
  with many concurrent workers and/or slow initializing workers.
* `commands` - A single directory, an array of directories, or a comma-delimited list of directories
  may be provided to auto-register commands found in the provided folders that match the ".js"
  extension. If the module exposes the "id" property, that will be the name of the command,
  otherwise the filename (minus the extension) will be used as the name of the command. If relative
  paths are provided, they will be resolved from process.cwd().
* `master` - An optional module to execute for the master process only, once ```start``` has been completed.
* `proxy` - Optional path to a JSON config file to enable Proxy Support.
* `workerGid` - Group ID to assign to child worker processes (recommended `nobody`).
* `workerUid` - User ID to assign to child worker processes (recommended `nobody`).


## Console & REST API

A DPS Cluster Service has two interfaces, the console (stdio), and an HTTP REST API. The two
interfaces are treated identical, as console input/output is piped over the REST API. The
reason for the piping is that a DPS Cluster Service is intentionally designed to only
support one instance of the given service running at any one time, and the port binding
is the resource constraint. This allows secondary services to act as console-only
interfaces as they pipe all input/output over HTTP to the already running service
that owns the port. This flow enables the CLI to background processes.
The REST API is locked to a "accessKey" expected in the query string. The console
automatically passes this key to the REST API, but for external REST API access,
the key will need to be known.

	{ host: "localhost", port: 11987, accessKey: "lksjdf982734" }

Invoking the REST interface directly would look something like:

	curl -d "" "http://localhost:11987/cli?cmd=help&accessKey=lksjdf982734"

Or better yet, use the ```run``` option to do the work for you:

	cservice "help" --accessKey "lksjdf982734"
	// same as
	cservice --run "help" --accessKey "lksjdf982734"



## Cluster Commands

While a Cluster Service may provide its own custom commands, below are provided out-of-the-box.
Commands may be disabled by overriding them.

* `start workerPath [cwd] { [timeout:60000] }` - Gracefully start service, one worker at a time.
* `restart all|pid { [timeout:60000] }` - Gracefully restart service, waiting up to timeout before terminating workers.
* `shutdown all|pid { [timeout:60000] }` - Gracefully shutdown service, waiting up to timeout before terminating workers.
* `exit now` - Forcefully exits the service.
* `help [cmd]` - Get help.
* `upgrade all|pid workerPath { [cwd] [timeout:60000] }` - Gracefully upgrade service, one worker at a time. (continuous deployment support).
* `workers` - Returns list of active worker processes.
* `health` - Returns health of service. Can be overidden by service to expose app-specific data.
* `info` - Returns summary of process & workers.



## Commands & Events

Creating custom, or overriding commands and events is as simple as:

  cservice "server.js" --commands "./commands,../some_more_commands"

Or if you prefer to manually do so via code:

	var cservice = require("cluster-service");
	cservice.on("custom", function(evt, cb, arg1, arg2) { // "custom" command
		// can also fire custom events
		cservice.trigger("on.custom.complete", 1, 2, 3);
	};

	cservice.on("test", function(evt, cb, testScript, timeout) { // we're overriding the "test" command
		// arguments
		// do something, no callback required (events may optionally be triggered)
	};

	// can also issue commands programatically
	cservice.trigger("custom", function(err) { /* my callback */ }, "arg1value", "arg2value");


## Cluster Events

Events are emitted to interested parties.

* `workerStart (pid, reason)` - Upon exit of any worker process, the process id of the exited worker. Reasons include: "start", "restart", "failure", and "upgrade".
* `workerExit (pid, reason)` - Upon start of any worker process. Reasons include: "start", "restart", "failure", and "upgrade".



## Async Support

While web servers are automatically wired up and do not require async logic (as of v1.0), if
your service requires any other asynchronous initialization code before being ready, this
is how it can be done.

Have the worker inform the master once it is actually ready:

	// worker.js
	require("cluster-service").workerReady(false); // we're NOT ready!
	setTimeout(funtion() {
		// dumb example of async support
		require("cluster-service").workerReady(); // we're ready!
	}, 1000);

Additionally, a worker may optionally perform cleanup tasks prior to exit, via:

	// worker.js
	require("cluster-service").workerReady({
		onWorkerStop: function() {
			// lets clean this place up
			process.exit(); // we're responsible for exiting if we register this cb
		}
	});



## Access Control

Commands may be granted "inproc" (high risk), "local" (low risk), or "remote" (no risk). Setting
access control can be done within the command, like so:

```javascript
// exit.js
module.exports.control = function(){
	return "local";
};
```

Or may be overriden at runtime via:

```javascript
// server.js
require("cluster-service").control({ "exit": "local" });
```

## Proxy Support

Proxy mode specifically caters to Web Servers that you want to enable automatic
versioning of your service. Any version requested (via `versionHeader`) that is
not yet loaded will automatically have a worker process spun up with the new
version, and after ready, the proxy will route to that worker.

Every version of your app *must* adhere to the `PROXY_PORT` environment
variable like so:

	require("http").createServer(function(req, res) {
	  res.writeHead(200);
	  res.end("Hello world!");
	}).listen(process.env.PROXY_PORT || 3000 /* port to use when not running in proxy mode */);

### Proxy Options

* `versionPath` (default: same directory as proxy JSON config) - Can override
  to point to a new version folder.
* `defaultVersion` - The version (folder name) that is currently active/live.
  If you do not initially set this option, making a request to the Proxy without
	a `versionHeader` will result in a 404 (Not Found) since there is no active/live
	version.
  Upgrades will automatically update this option to the latest upgraded version.
* `versionHeader` (default: `x-version`) - HTTP Header to use when determining
  non-default version to route to.
* `workerFilename` (default: `worker.js`) - Filename of worker file.
* `bindings` (default: `[{ port: 80, workerCount: 2 }]`) - An array of `Proxy Bindings`.
* `versionPorts` (default: `11000-12000`) - Reserved port range that can be used to
  assign ports to different App versions via `PROXY_PORT`.
* `nonDefaultWorkerCount` (default: 1) - If a version is requested that is not
  a default version, this will be the number of worker processes dedicated to
	that version.
* `nonDefaultWorkerIdleTime` (default: 3600) - The number of seconds of inactivity
  before a non-default version will have its workers shut down.

### Proxy Bindings

Binding options:

* `port` - Proxy port to bind to.
* `workerCount` (default: 2) - Number of worker processes to use for this
  binding. Typically more than 2 is unnecessary for a proxy, and less than 2
	is a potential failure point if a proxy worker ever goes down.
* `tlsOptions` - TLS Options if binding for HTTPS.
  * `key` - Filename that contains the Certificate Key.
	* `cert` - Filename that contains the Certificate.
	* `pem` - Filename that contains the Certificate PEM if applicable.

A full list of TLS Options: https://nodejs.org/api/tls.html#tls_tls_createserver_options_secureconnectionlistener

### Proxy Commands

Work like any other `Cluster Commands`.

* `proxy start configPath` - Start the proxy using the provided JSON config file.
* `proxy stop` - Shutdown the proxy service.
* `proxy version workerVersion workerCount` - Set a given App version to the
  desired number of worker processes. If the version is not already running,
	it will be started. If 2 workers for the version are already running, and you
	request 4, 2 more will be started. If 4 workers for the version are already
	running, and you request 2, 2 will be stopped.
* `proxy promote workerVersion workerCount` - Works identical to the
  `proxy version` command, except this will flag the version as active/live,
	resulting in the Proxy Config file being updated with the new `defaultVersion`.
* `proxy info` - Fetch information about the proxy service.



## Tests & Code Coverage

Download and install:

	git clone https://github.com/godaddy/node-cluster-service.git
	cd node-cluster-service
	npm install

Now test:

	npm test

View code coverage in any browser:

	coverage/lcov-report/index.html


## Change Log

[Change Log](https://github.com/godaddy/node-cluster-service/blob/master/CHANGELOG.md)



## License

[MIT](https://github.com/godaddy/node-cluster-service/blob/master/LICENSE.txt)


CHANGELOG.md
## v2.1.4 - 9/21/2020

### Fixes:

- Dependency updates.

## v2.1.1 - 5/30/2018

### Fixes:

- Memory leak in message bus (thanks @jakepusateri !)- #78

## v2.1.0 - 9/1/2016

### Breaking:

- Updated to require Node v6 or later.

## v2.0.0 - 9/1/2016

### Features:

- High worker restart concurrency to speed up upgrades/restarts via `restartConcurrencyRatio`
- Proxy support with dynamic app versioning
  - Central proxy config file, promotions persisted to disk to survive service restarts
  - Isolated worker processes per version
  - On demand versioning via `x-version` header
  - Version promotion brings up worker count for desired version and brings down worker count of old version
- Worker process downgrade from root via `workerGid` & `workerUid`
- Backward compatibility with `v1.x`! Who does that with a major release?!

### Enhancements:

- Refactored communication between processes
- Worker processes may now trigger commands (same) and wait for response (new).
  The new message bus enables RPC and other flows like this between processes

### Fixes:

- Worker process.send crash

## v1.0.0 - 5/8/2014

### Features:

- #56. Ability to restart web server gracefully under load with no code
- No code required to enable net statistics

### Fixes:

- Prevent additional commands in CLI while in progress to resolve state issues
- Handle failed `start` command gracefully
- Handle failed `upgrade` command gracefully
- Fixed `start` command from preventing failed workers to auto-restart
- Fixed `restart` command from preventing failed workers to auto-restart
- Further streamlined management of state across multiple cservice instances
- Allow numeric accessKey's
- Fix crash if no command provided in REST call
- #57. Cannot use '--run' and '--config' together
- #58. Custom events from master & workers

## v0.10.0 - 5/2/2014

### Legacy:

- Dropped support for 'ready' flag, was causing support/complexity issues

### Features:

- Ability to start without worker
- Ability to support multiple versions of cservice through shared state
- Added example workers to `examples/`

## v0.9.0 - 2/17/2014

### Features:

- #51. Lockdown "help" command to "local" by default
- #49. Support for multiple keys within 'accessKey' option. (Undocumented,experimental)
- #54. Allow REST command to originate from body

### Fixes:

- #46. Expect API response even on "shutdown" or "exit"
- #52. processDetails not returned during message floods
- #53. Net Statistics doesn't work if using multiple instances of cservice

## v0.8.0 - 1/13/2014

### Features:

- #40. Net statistics support

## v0.7.0 - 1/3/2014

### Features:

- #39. Support for hidden(internal) commands
- #38. Option 'commands' support
- #37. Disable cli by default, unless run from command-line
- #30. Smarter async workerReady logic


lib/proxy.js
var cservice = require("../cluster-service");
var cluster = require("cluster");
var workersHelper = require("./workers");
var path = require("path");
var async = require("async");
var fs = require("fs");

module.exports = {
  start: start,
  stop: stop,
  version: version,
  promote: promote,
  updateProxyWorkers: updateProxyWorkers,
  getProxyWorkers: getProxyWorkers,
  getVersionWorkers: getVersionWorkers,
  info: info
};

function start(options, cb) {
  if (cluster.isWorker === true) {
    cservice.log("Proxy cannot be started from worker".warn);
    return cb && cb("Proxy cannot be started from worker");
  }

  var configPath = options.configPath || cservice.options.proxy;

  if (typeof configPath !== "string") {
    // disabled
    return cb && cb();
  }

  if (cservice.locals.proxy.enabled === true) {
    cservice.log("Proxy already running".warn);
    return cb && cb("Proxy already running");
  }

  cservice.locals.proxy.configPath = path.resolve(configPath);
  cservice.locals.proxy.options =
    JSON.parse(fs.readFileSync(cservice.locals.proxy.configPath))
  ;

  options = cservice.locals.proxy.options;

  if (Array.isArray(options.bindings) === false ||
    options.bindings.length === 0) {
    options.bindings = [{ port: 80, workerCount: 2 }]; // default
  }

  options.nonDefaultWorkerCount = options.nonDefaultWorkerCount || 1;
  options.nonDefaultWorkerIdleTime = options.nonDefaultWorkerIdleTime || 3600;

  cservice.locals.proxy.versionPath =
    options.versionPath || path.dirname(cservice.locals.proxy.configPath)
  ;
  cservice.locals.proxy.versionPath =
    path.resolve(cservice.locals.proxy.versionPath)
  ;
  cservice.locals.proxy.workerFilename = options.workerFilename || "worker.js";
  cservice.locals.proxy.versionHeader = options.versionHeader || "x-version";

  var portRange = (options.versionPorts || "11000-12000").split("-");
  cservice.locals.proxy.portRange = {
    min: parseInt(portRange[0]),
    max: parseInt(portRange[1])
  };
  cservice.locals.proxy.nextAvailablePortIndex = 0;

  var proxyWorkerTasks = options.bindings.map(function(b) {
    return function(cb) {
      var workerOptions = {
        type: "proxy", // proxy worker type
        worker: path.resolve(__dirname, "proxy-worker.js"),
        bindingInfo: JSON.stringify(b),
        versionPath: cservice.locals.proxy.versionPath,
        versionHeader: cservice.locals.proxy.versionHeader,
        workerFilename: cservice.locals.proxy.workerFilename,
        versions: JSON.stringify(cservice.locals.proxy.versions),
        count: b.workerCount || 2
      };
      cservice.newWorker(workerOptions, cb);
    };
  });

  cservice.locals.proxy.refreshnessTimer =
    setInterval(checkVersionsForFreshness, 10000)
  ;
  cservice.locals.proxy.refreshnessTimer.unref();

  cservice.locals.proxy.enabled = true;

  async.parallel(proxyWorkerTasks, function (err) {
    var portArr = options.bindings.map(function(b) {
      return b.port.toString().data;
    });

    if (err) {
      cservice.error("Proxy failed to run on ports ".error +
        portArr.join(",".info) + " with error ".error + err.toString().data);
      return cb && cb(err);
    }

    cservice.log("Proxy running on ports ".info + portArr.join(",".info));

    if (!cservice.locals.proxy.options.defaultVersion) {
      // no current version

      return cb && cb();
    }

    version(cservice.locals.proxy.options.defaultVersion,
      { workerCount: cservice.locals.options.workerCount },
      function (err, version) {
      if (err) {
        cservice.error("Proxy failed to run on ports ".error +
          portArr.join(",".info) + " with error ".error + err.toString().data
        );
        return cb && cb(err);
      }

      return cb && cb();
    });
  });
}

function stop(cb) {
  if (cluster.isWorker === true) {
    cservice.log("Proxy cannot be stopped from worker".warn);
    return cb && cb("Proxy cannot be stopped from worker");
  }

  if (typeof cservice.locals.proxy.configPath !== "string") {
    // disabled
    return cb && cb();
  }

  if (cservice.locals.proxy.enabled === false) {
    cservice.log("Proxy not running".warn);
    return cb && cb("Proxy not running");
  }

  clearInterval(cservice.locals.proxy.refreshnessTimer);
  cservice.locals.proxy.refreshnessTimer = null;

  // now lets trigger a shutdown
  cservice.trigger("shutdown", function(err, result) {
    cservice.locals.proxy.enabled = false;
    return cb && cb();
  }, "all");
}

function version(versionStr, options, cb) {
  if (cluster.isWorker === true) {
    cservice.log("Proxy cannot invoke 'version' from worker".warn);
    return cb && cb("Proxy cannot invoke 'version' from worker");
  }

  options = options || {};
  if (isNaN(options.workerCount) === true) {
    options.workerCount =
      (versionStr === cservice.locals.proxy.options.defaultVersion)
      ? cservice.locals.options.workerCount
      : cservice.locals.proxy.options.nonDefaultWorkerCount
    ;
  }

  // detect current version worker count
  var currentVersionWorkers = getVersionWorkers(versionStr);

  // determine worker delta from desired count and actual count
  var workerCountDelta = options.workerCount - currentVersionWorkers.length;

  // get existing version listing
  var v = cservice.locals.proxy.versions[versionStr];

  if (v) {
    // update version lastAccess
    v.lastAccess = Date.now();
  }

  // if version worker count is already current, nothing more to do
  if (workerCountDelta === 0) {
    return cb && cb();
  }

  if (workerCountDelta < 0) {
    // if desired version count is less than current, reduce worker count
    var workersToShutdown = currentVersionWorkers.length - options.workerCount;
    var shutdownTasks = [];
    for (var workerToShutdown = 0; workerToShutdown < workersToShutdown;
      workerToShutdown++) {
      shutdownTasks.push(
        getWorkerShutdownTask(
          currentVersionWorkers[workerToShutdown],
          options.reason || "proxy version"
        )
      );
    }

    // shutdown all at once
    async.parallel(shutdownTasks, function (err, results) {
      return cb && cb(err);
    });

    return;
  }

  // if desired version count is more than current, spin up new workers

  var workerPath = path.resolve(cservice.locals.proxy.versionPath,
    versionStr, cservice.locals.proxy.workerFilename
  );

  // use existing port if available, otherwise allocate a new one
  var versionPort = (v && v.port) || getNextAvailablePort();

  cservice.trigger("start", function (err, result) {
    return cb && cb(err, result);
  }, workerPath, {
    count: workerCountDelta,
    version: versionStr,
    PROXY_PORT: versionPort
  });
}

function getWorkerShutdownTask(worker, reason) {
  return function(cb) {
    worker.cservice.reason = reason;
    cservice.trigger("shutdown", cb, worker.process.pid);
  };
}

function getNextAvailablePort() {
  // always continue where we left off from the last time
  // we fetched an available port.
  // this generally will allow us to return in O(1),
  // unless there are tons of active versions.
  var totalPorts = cservice.locals.proxy.portRange.max
    - cservice.locals.proxy.portRange.min
  ;
  for (var i = 0; i < totalPorts; i++) {
    var port = cservice.locals.proxy.portRange.min
      + ((cservice.locals.proxy.nextAvailablePortIndex + i) % totalPorts);
    if (isPortInUse(port) === false) {
      cservice.locals.proxy.nextAvailablePortIndex = ((i + 1) % totalPorts);
      return port;
    }
  }

  throw new Error(
    "All proxy ports have been used up! Try increasing range of " +
    "`proxy.versionPorts` or reducing `proxy.nonDefaultWorkerIdleTime`.");
}

function isPortInUse(port) {
  for (var k in cservice.locals.proxy.versions) {
    if (cservice.locals.proxy.versions.hasOwnProperty(k) === false) {
      continue; // ignore
    }
    if (cservice.locals.proxy.versions[k].port === port) {
      return true; // NOT available
    }
  }

  // if we get this far, we're OK to use port
  return false;
}

function promote(versionStr, options, cb) {
  if (cluster.isWorker === true) {
    cservice.log("Proxy cannot invoke 'promote' from worker".warn);
    return cb && cb("Proxy cannot invoke 'promote' from worker");
  }

  options = options || {};
  options.workerCount = options.workerCount ||
    cservice.locals.options.workerCount
  ;

  var oldVersion = cservice.locals.proxy.versions[
    cservice.locals.proxy.options.defaultVersion
  ];

  // set to-be-promoted version to desired worker count
  version(versionStr, options, function (err) {
    if (err) {
      // pass failure on
      return cb && cb(err);
    }

    // persist to-be-promoted version
    cservice.locals.proxy.options.defaultVersion = versionStr;
    fs.writeFile(cservice.locals.proxy.configPath,
      JSON.stringify(cservice.locals.proxy.options, null, "  "), function(err) {
      if (err) {
        // pass failure on
        cservice.error("Failed to proxy promote version".error +
          versionStr.info
        );
        return cb && cb(err);
      }

      // notify proxy-workers of promoted version
      updateProxyWorkers();

      cservice.log("Proxy promoted version ".success +
        versionStr.info +
        " successfully".success
      );

      // bring previously promoted version down to
      // `nonDefaultWorkerCount` workers, but no need to wait for callback
      if (oldVersion && oldVersion.name !== versionStr) {
        version(oldVersion.name,
          {
            workerCount: cservice.locals.proxy.options.nonDefaultWorkerCount,
            reason: "proxy demote"
          }, function(err) {
            if (err) {
              return cservice.error("Failed to proxy demote version".error +
                oldVersion.name.info
              );
            }

            cservice.log("Proxy demoted old version ".success +
              oldVersion.name.info +
              " successfully".success
            );
          }
        );
      }

      if (cb) {
        setImmediate(cb);
      }
    });
  });
}

function updateProxyWorkers() {
  var msg = cservice.msgBus.createMessage("proxyVersions", {
    versions: cservice.locals.proxy.versions,
    defaultVersion: cservice.locals.proxy.options.defaultVersion
  });
  getProxyWorkers().forEach(function(worker) {
    worker.send(msg);
  });
}

function info(cb) {
  if (cluster.isWorker === true) {
    cservice.log("Proxy cannot invoke 'info' from worker".warn);
    return cb && cb("Proxy cannot invoke 'info' from worker");
  }

  var now = Date.now();
  var proxyWorkers = getProxyWorkers().map(function(worker) {
    var bindingInfo = JSON.parse(worker.cservice.bindingInfo);
    return {
      port: bindingInfo.port,
      ssl: typeof bindingInfo.tlsOptions === "object"
    };
  });
  var versionWorkers = getVersionWorkers().map(function(worker) {
    var versionInfo = cservice.locals.proxy.versions[worker.cservice.version];
    return {
      worker: worker.cservice.worker,
      version: worker.cservice.version,
      lastAccess: versionInfo ?
        Math.round((now - versionInfo.lastAccess) / 1000) : "?"
    };
  });

  cb(null, {
    versionPath: cservice.locals.proxy.versionPath,
    workerFilename: cservice.locals.proxy.workerFilename,
    portRange: cservice.locals.proxy.portRange,
    options: cservice.locals.proxy.options,
    proxyWorkers: proxyWorkers,
    versionWorkers: versionWorkers
  });

}

function getProxyWorkers() {
  return cservice.workers.filter(function(worker) {
    return worker.cservice.type === "proxy";
  });
}

function getVersionWorkers(explicitVersion) {
  return cservice.workers.filter(function(worker) {
    var result =
      typeof worker.cservice.version === "string" &&
      (!explicitVersion || worker.cservice.version === explicitVersion)
    ;
    return result;
  });
}

function isVersionRunning(versionStr, cb) {
  if (!(versionStr in cservice.locals.proxy.versions)) {
    return false; // version not available
  }

  // are any worker processes running desired version?
  var workers = cservice.locals.workerProcesses;

  for (var i = 0; i < workers.length; i++) {
    var worker = workers[i];
    var pid = worker.process.pid;
  }

  return false; // no workers running desired version
}

function checkVersionsForFreshness() {
  var now = Date.now();
  for (var k in cservice.locals.proxy.versions) {
    if (
      // live version is exempt
      k === cservice.locals.proxy.options.defaultVersion ||
      // verify a valid version
      !cservice.locals.proxy.versions.hasOwnProperty(k)) {
      continue; // skip
    }
    var v = cservice.locals.proxy.versions[k];
    var diff = (now - v.lastAccess) / 1000; // seconds
    if (diff < cservice.locals.proxy.options.nonDefaultWorkerIdleTime) {
      continue; // all OK
    }

    cservice.log("Proxy version ".warn + k.info +
      " shutting down due to inactivity".warn
    );

    // kill all the things
    version(k, { workerCount: 0 });
  }
}
